raft/heartbeat_manager: race condition on _fstats

joe-redpanda opened the pull request:
hearbeat manager accesses two sets of collections in gathering heartbeat data for a given round of heartbeat RPCs.

1. _consensus_groups which is owned by hbm, and protected by a lock for the duration of gathering heartbeats
2. consensus::_fstats which is NOT owned by hbm and not protected by any locking

heartbeat manager uses ssx::async_for_each_counter to gather heartbeat data because there are feasibly enough raft groups, and followers in those raft groups to either stall the reactor and or meaningfully increase e2e latency if all of these were computed in one synchronous block.

As previously written, hbm may suspend in the middle of an iteration over an _fstats object. If consensus processes a reconfiguration while hbm's iteration is suspended, fstats will be mutated and the in-use iterator will be invalidated.

This PR changes the loop over _fstats to be synchronous s.t. _fstats may no longer mutate while in use by hbm. To avoid reactor stalls and latency concerns, the outer loop is now asynchronous and set to offer a yield every 30 iterations. We make the assumption that each raft group will have approximately four followers, thus approximately preserving the prior logic of one yield every 100 followers.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

dotnwat commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -54,103 +49,127 @@ heartbeat_manager::follower_request_meta::follower_request_meta(
-                  reconnect_nodes.insert(id.id());
-              }
-          });
+        // the actual work for creating heartbeats will happen async
+        co_await fetch_heartbeats_for_raft_group(
+          raft_group, pending_beats, reconnect_nodes, last_heartbeat);
```
it seems like the original inner loop was replaced with this

> co_await ssx::async_for_each_counter(counter, r->_fstats.begin(), r->_fstats.end(),

to avoid reactor stalls by introducing >= 0 scheduler yields (zero if the set being operated on is small). However, it the change in this PR introduces a single guaranteed preemption check `co_await ss::maybe_yield();` plus  a non-interruptible inner loop over `raft_group->_fstats`. So it looks like now we have <= 1 scheduler yields. Is this correct? It seems like the same reactor stall concern would still exist?

mmaslankaprv replied:
Exactly, i think we need to find different solution for this problem, the solution should not involve locking and should be asynchronous

joe-redpanda replied:
My thought was that followers should be at max with _raft0 on a large cluster which I think reasonably maxes out at 100ishish nodes

where the number of raft groups could vary substantially more, so the more predictably sized workload is "for each follower".

The inner loop also isn't terribly CPU heavy (to my eyes, open to being wrong): theres no heavy duty serialization ongoing.

Solution either way would be to offer a 'maybe_yield' within the loop, thoughts?

dotnwat replied:
> Solution either way would be to offer a 'maybe_yield' within the loop, thoughts?

my understanding is that the issue was accounting anomolies that occured when there were scheduling points in that loop. if that is correct, then adding a maybe_yield would seem like it would reintroduce those scheduling points and we'd back to having the same issue?

> The inner loop also isn't terribly CPU heavy (to my eyes, open to being wrong): theres no heavy duty serialization ongoing.

I think it's a valid point. The original PR by @mmaslankaprv seems to suggest that no reactor stall was actually observed. Maybe the right thing to do here is a simple benchmark of this loop (or something like this loop) to see when we might reach a cost that would cause a reactor stall. Then if the number is high enough, we can just ignore it?

https://github.com/redpanda-data/redpanda/commit/86aebfe513a76cb757f62ed60bbe45915cddb4cc

joe-redpanda replied:
Thats fine so long as we have a copy of the list, the issue is that we lose atomicity across a scheduling boundary

I do also have concerns that the more scheduling points we add, the more likely it is that two heartbeat requests start overlapping in proper time as the creation of the request gets smeared across the task queue.

I don't think that should cause issues but I can't imagine it does anything good either

dotnwat replied:
tbh i'm not sure what is being proposed anymore.

> we lose atomicity across a scheduling boundar

is the implication that stats are inconsistent, or something worse is happening?

mmaslankaprv replied:
The copy is not needed as the `_consensus_groups` set is protected by the mutex, the only part which is concerning is iterating over the `_fstats`. The loop is executed all the time in Redpanda and using `async_for_each` is an important optimization. 

My suggestion is: either check if the map changed while iterating over it or create dedicated `async_for_each` like function that would never yield in the loop but rather than that do it after.

dotnwat replied:
is the issue with using invalid iterators during iteration, or something like statistics are inconsistent?

joe-redpanda replied:
The issue is with using invalid iterators during iteration.

https://github.com/redpanda-data/redpanda/pull/27070

I added a cost calculation to the for_each* family of algos

I'm planning to follow the test examplse and set a cost function determined by the number of followers in a given _fstats and putting the async_for_each on the outer loop

We'll yield when its safe (after we're done using an _fstats), and only after seeing 100 followers

dotnwat replied:
Sorry, I still don't understand.

> Heartbeat manager races reconfigurations for follower_metadata contained
in consensus::_fstats.

From the coverletter, it sounds like the problem is a race condition. But I can't seem to figure out the details.

joe-redpanda replied:
hb_m builds heartbeats (currently) with something like

```
for(consensus* raft_group : _raft_groups)
    async for(auto follower : raft_group->_fstats)
        ... // do some stuff
```

we have a lock on _raft_groups
we don't have a lock on _fstats

async_for_each_counter reserves the right to suspend at any given point in its iteration
so the 'async for(auto follower : raft_group->_fstats)' can suspend in the middle of an _fstats iteration

while suspended _fstats can be changed by consensus in a reconfiguration, and then our iterator is invalid|


so, one of the possible solutions is to just... not yield while using _fstats

dotnwat replied:
please update the coverletter and commit message to explain what is going on

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4938e76a-2e38-4245-9fc3-8b30b2d36da5-->
<!--TEST RESULTS-->
<!--9d8b3df0-0a14-4c69-b13d-ae5cd27b1362-->
<!--TEST RESULTS-->
<!--a4cdcc84-0432-412c-a4f1-4c2e1f4d97a1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69697">build#69697</a></summary>


| test_class               | test_method             | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest          | test_maintenance_sticky | {"use_rpk": true}                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/69697#019843ec-3477-450b-a3cb-7ac4553dc4f9 | FLAKY         | 20/21    | upstream reliability is '99.44444444444444'. current run reliability is '95.23809523809523'. drift is 4.20635 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69697#019843ec-3477-4f9d-bd3a-1f91e48ca529 | FLAKY         | 20/21    | upstream reliability is '99.65870307167235'. current run reliability is '95.23809523809523'. drift is 4.42061 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70045">build#70045</a></summary>


| test_class             | test_method                          | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------------|:-------------------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest  | test_creating_and_listing_migrations | null                                                | integration | https://buildkite.com/redpanda/redpanda/builds/70045#019862fc-40de-4fc6-a93e-f0f37474a94d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| DatalakeThrottlingTest | test_basic_throttling                | {"catalog_type": "nessie", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/70045#019862fc-40db-4c00-a2e4-46da6382f27e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70098">build#70098</a></summary>


| test_class               | test_method            | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupTest        | test_group_lag_metrics | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/70098#0198670a-3308-4a86-90a6-c57688e54b1e | FLAKY         | 20/21    | upstream reliability is '99.25742574257426'. current run reliability is '95.23809523809523'. drift is 4.01933 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations   | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | integration | https://buildkite.com/redpanda/redpanda/builds/70098#0198670a-3307-47d6-85a2-3adf225d21d8 | FLAKY         | 20/21    | upstream reliability is '99.39577039274926'. current run reliability is '95.23809523809523'. drift is 4.15768 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

bharathv commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -54,103 +56,126 @@ heartbeat_manager::follower_request_meta::follower_request_meta(
-          });
-    }
+    // copy raft groups to
+    // 1. avoid maintaining a lock in a timed bg fiber
```
(1) is not included in this PR, right?

bharathv replied:
q: After going through the full diff, I'm wondering if this copy is really needed to fix the proposed problem, can't we just do

```
   co_await ssx::async_for_each_counter<loop_traits>(
      counter,
      _consensus_groups.begin(),
      _consensus_groups.end(),
      [this, &pending_beats, &reconnect_nodes, last_heartbeat](
        ss::lw_shared_ptr<consensus>& raft_group) {
          fetch_heartbeats_for_raft_group(
            raft_group, pending_beats, reconnect_nodes, last_heartbeat);
      });
```

under the mutex? We were already holding that mutex before and we don't see any issues due to that contention, so why fix that?

joe-redpanda replied:
Yeah, thats a forgotten relic of an old solution. Removed.

bharathv commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -54,103 +56,126 @@ heartbeat_manager::follower_request_meta::follower_request_meta(
+    ssx::async_counter counter{};
+    co_await ssx::async_for_each_counter<loop_traits>(
+      counter,
+      _consensus_groups.begin(),
```
did you mean `snapped_raft_groups`?

joe-redpanda replied:
mistake, that was meant to be removed given that we're protected by _lock

bharathv commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -54,103 +56,126 @@ heartbeat_manager::follower_request_meta::follower_request_meta(
+    // copy raft groups to
+    // 1. avoid maintaining a lock in a timed bg fiber
+    // 2. ensure the collection cannot change while iterating
+    consensus_set snapped_raft_groups = _consensus_groups;
```
How expensive is this? Its a set\<ptr\> and can be as big as # of partitions on a shard, so ~ 5000 ptrs? Every 150ms, is that correct?

mmaslankaprv commented:
@joe-redpanda can you explain in cover letter that the `_fstats` could be modified by the `consensus` while iterating over them to collect heartbeat.

bharathv reviewed (approved):
lgtm pending cover letter update.

joe-redpanda committed:
raft/heartbeat_manager: race condition on _fstats

hearbeat manager accesses two sets of collections in gathering
heartbeat data for a given round of heartbeat RPCs.

1. _consensus_groups which is owned by hbm, and protected by a lock for
   the duration of gathering heartbeats
2. consensus::_fstats which is NOT owned by hbm and not protected by any
   locking

heartbeat manager uses ssx::async_for_each_counter to gather heartbeat
data because there are feasibly enough raft groups, and followers in
those raft groups to either stall the reactor and or meaningfully
increase e2e latency if all of these were computed in one synchronous
block.

As previously written, hbm may suspend in the middle of an iteration
over an _fstats object. If consensus processes a reconfiguration
while hbm's iteration is suspended, fstats will be mutated and the
in-use iterator will be invalidated.

This PR changes the loop over _fstats to be synchronous s.t. _fstats may
no longer mutate while in use by hbm. To avoid reactor stalls and
latency concerns, the outer loop is now asynchronous and set to offer a
yield every 30 iterations. We make the assumption that each raft group
will have approximately four followers, thus approximately preserving
the prior logic of one yield every 100 followers.

dotnwat reviewed (dismissed):
> Heartbeat manager races reconfigurations for follower_metadata contained
in consensus::_fstats. This PR reshuffles the access pattern s.t. the
usage of _fstats will only happen in a sync codespan and be atomic

changes lgtm, but the commit messsage needs to be updated to provide more and clearer context about the change. see here for tips https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#content

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Oops! Something went wrong.

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16756570391).
